runtime/js-compute-runtime/builtins/crypto-algorithm.cpp

```diff
@@ -755,12 +827,167 @@ JSObject *CryptoAlgorithmHMAC_Sign_Verify::toObject(JSContext *cx) {
+
+  auto digest = digestOption.value();
+
+  // 4. Let d be the ECDSA private key associated with key.
+  #pragma clang diagnostic push
+  #pragma clang diagnostic ignored "-Wdeprecated-declarations"
+  const EC_KEY * ecKey = EVP_PKEY_get0_EC_KEY(CryptoKey::key(key));
+  #pragma clang diagnostic pop
+  if (!ecKey) {
+    DOMException::raise(cx, "SubtleCrypto.verify: failed to verify", "OperationError");
+    return nullptr;
+  }
+
+  // 5. Let params be the EC domain parameters associated with key.
+  #pragma clang diagnostic push
+  #pragma clang diagnostic ignored "-Wdeprecated-declarations"
+  auto sig = ECDSA_do_sign(digest.data(), digest.size(), const_cast<EC_KEY*>(ecKey));
+  #pragma clang diagnostic pop
+  if (!sig) {
+    DOMException::raise(cx, "SubtleCrypto.verify: failed to verify", "OperationError");
+    return nullptr;
+  }
+
+  // 6. If the namedCurve attribute of the [[algorithm]] internal slot of key is "P-256", "P-384" or "P-521":
+  //         Perform the ECDSA signing process, as specified in [RFC6090], Section 5.4, with M as the message, using params as the EC domain parameters, and with d as the private key.
+  //         Let r and s be the pair of integers resulting from performing the ECDSA signing process.
+  //         Let result be an empty byte sequence.
+  //         Let n be the smallest integer such that n * 8 is greater than the logarithm to base 2 of the order of the base point of the elliptic curve identified by params.
+  //         Convert r to an octet string of length n and append this sequence of bytes to result.
+  //         Convert s to an octet string of length n and append this sequence of bytes to result.
+  // Otherwise, the namedCurve attribute of the [[algorithm]] internal slot of key is a value specified in an applicable specification:
+  //     Perform the ECDSA signature steps specified in that specification, passing in M, params and d and resulting in result.
+  const BIGNUM* r;
+  const BIGNUM* s;
+  ECDSA_SIG_get0(sig, &r, &s);
+  auto keySize = curveSize(cx, key);
+  if (keySize.isErr()) {
+    return nullptr;
+  }
+
+  size_t keySizeInBytes = numBitsToBytes(keySize.unwrap());
+
+  auto rBytesAndSize = convertToBytesExpand(cx, r, keySizeInBytes);
+  if (!rBytesAndSize.has_value()) {}
+  auto rBytes = std::move(rBytesAndSize.value().first);
+  auto rBytesSize = std::move(rBytesAndSize.value().second);
+
+  auto sBytesAndSize = convertToBytesExpand(cx, s, keySizeInBytes);
+  if (!sBytesAndSize.has_value()) {}
+  auto sBytes = std::move(sBytesAndSize.value().first);
+  auto sBytesSize = std::move(sBytesAndSize.value().second);
+
+  auto resultSize = rBytesSize + sBytesSize;
+  mozilla::UniquePtr<uint8_t[], JS::FreePolicy> result{
+      static_cast<uint8_t *>(JS_malloc(cx, resultSize))};
+
+  std::memcpy(result.get(), rBytes.get(), rBytesSize);
+  std::memcpy(result.get() + rBytesSize, sBytes.get(), sBytesSize);
+  static_cast<void>(rBytes.release());
+  static_cast<void>(sBytes.release());
```

Comment:
@elliottt I feel like this part could likely be improved but not too sure what I'd do different

Reply:
You could modify `convertToBytesExpand` to accept the output pointer to write to, and do the realloc there. However, I think what you've done is totally reasonable, and I wouldn't worry about being that careful with memory for a function that shouldn't be called on a hot path.

One change that I would make is to change the uses of `std::move` to the `.get()` method on the `UniquePtr`, as that will give you access to the pointer for copying. That way you can remove the `.release()` calls, as the memory will still be managed by the `UniquePtr`s.

Reply:
Good idea, I applied what I think you mean in this commit -- [`c78b288` (#667)](https://github.com/fastly/js-compute-runtime/pull/667/commits/c78b288f4f36aaf382407f66ebc2ab763ec4ac86)